librz/analysis/arch/arm/arm_il64.c

```diff
@@ -0,0 +1,2707 @@
+	cs_arm64_op *op = &insn->detail->arm64.operands[1];
+	ut32 shift = op->shift.type == ARM64_SFT_LSL ? op->shift.value : 0;
+	return write_reg(REGID(0), UN(REGBITS(0), ~(((ut64)op->imm) << shift)));
+}
+
+/**
+ * Capstone: ARM64_INS_MSR
+ * ARM: msr
+ */
+static RzILOpEffect *msr(cs_insn *insn) {
+	cs_arm64_op *op = &insn->detail->arm64.operands[0];
+#if CS_API_MAJOR > 4
+	if (op->type != ARM64_OP_SYS || op->sys != ARM64_SYSREG_NZCV) {
+		return NULL;
+	}
+#else
+	if (op->type != ARM64_OP_REG_MSR || op->reg != 0xda10) {
+		return NULL;
+	}
+#endif
+	ut32 bits = 0;
+	RzILOpBitVector *val = ARG(1, &bits);
+	if (!val) {
+		return NULL;
+	}
+	return SEQ4(
+		SETG("nf", INV(IS_ZERO(LOGAND(val, UN(bits, 1ull << 31))))),
+		SETG("zf", INV(IS_ZERO(LOGAND(DUP(val), UN(bits, 1ull << 30))))),
+		SETG("cf", INV(IS_ZERO(LOGAND(DUP(val), UN(bits, 1ull << 29))))),
+		SETG("vf", INV(IS_ZERO(LOGAND(DUP(val), UN(bits, 1ull << 28))))));
+}
+
+#if CS_API_MAJOR > 4
+/**
+ * Capstone: ARM64_INS_RMIF
+ * ARM: rmif
+ */
+static RzILOpEffect *rmif(cs_insn *insn) {
+	if (!ISIMM(1) || !ISIMM(2)) {
+		return NULL;
+	}
+	ut32 bits = 64;
+	RzILOpBitVector *val = ARG(0, &bits);
+	if (!val) {
+		return NULL;
+	}
+	ut64 lsb = IMM(1);
+	ut64 mask = IMM(2);
+	RzILOpEffect *eff = NULL;
+	const char *flags[] = { "vf", "cf", "zf", "nf" };
+	for (size_t i = 0; i < RZ_ARRAY_SIZE(flags); i++) {
+		if (!(mask & (1ull << i))) {
+			continue;
+		}
+		if (eff) {
+			val = DUP(val);
+		}
+		RzILOpEffect *set = SETG(flags[i], INV(IS_ZERO(LOGAND(val, UN(bits, 1ull << ((i + lsb) % 64))))));
+		eff = eff ? SEQ2(set, eff) : set;
+	}
+	if (!eff) {
+		rz_il_op_pure_free(val);
+	}
+	return eff ? eff : NOP;
+}
+#endif
+
+/**
+ * Capstone: ARM64_INS_SBFX, ARM64_INS_SBFIZ, ARM64_INS_UBFX, ARM64_INS_UBFIZ
+ * ARM: sbfx, sbfiz, ubfx, ubfiz
+ */
+static RzILOpEffect *sbfx(cs_insn *insn) {
+	if (!ISREG(0) || !ISIMM(2) || !ISIMM(3)) {
+		return NULL;
+	}
+	ut32 bits = REGBITS(0);
+	RzILOpBitVector *src = ARG(1, &bits);
+	if (!src) {
+		return NULL;
+	}
+	ut64 lsb = IMM(2);
+	ut64 width = IMM(3);
+	RzILOpBitVector *res;
+	if (insn->id == ARM64_INS_SBFIZ || insn->id == ARM64_INS_UBFIZ) {
+		res = SHIFTL0(UNSIGNED(width + lsb, src), UN(6, lsb));
+	} else {
+		// ARM64_INS_SBFX, ARM64_INS_UBFX
+		res = UNSIGNED(width, SHIFTR0(src, UN(6, lsb)));
+	}
+	bool is_signed = insn->id == ARM64_INS_SBFX || insn->id == ARM64_INS_SBFIZ;
+	res = LET("res", res, is_signed ? SIGNED(bits, VARLP("res")) : UNSIGNED(bits, VARLP("res")));
+	return write_reg(REGID(0), res);
+}
+
+/**
+ * Capstone: ARM64_INS_MRS
+ * ARM: mrs
+ */
+static RzILOpEffect *mrs(cs_insn *insn) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	cs_arm64_op *op = &insn->detail->arm64.operands[1];
+#if CS_API_MAJOR > 4
+	if (op->type != ARM64_OP_SYS || op->sys != ARM64_SYSREG_NZCV) {
+		return NULL;
+	}
+#else
+	if (op->type != ARM64_OP_REG_MRS || op->reg != 0xda10) {
```

Comment:
wasn't simpler to define `ARM64_SYSREG_NZCV` this on top of the file if `CS_API_MAJOR < 5` ?

Reply:
That would suggest that it's a stable constant, while it is rather just observation-based, which is the best we can do for cs4.